13-TeV-examples/uproot_python/HZZAnalysis.ipynb

```diff
@@ -444,63 +460,58 @@
-    "                                                            data.scaleFactor_PILEUP,\n",
-    "                                                            data.scaleFactor_ELE,\n",
-    "                                                            data.scaleFactor_MUON,\n",
-    "                                                            data.scaleFactor_LepTRIGGER)\n",
-    "\n",
-    "        # cut on lepton charge using the function cut_lep_charge defined above\n",
-    "        fail = data[ np.vectorize(cut_lep_charge)(data.lep_charge) ].index\n",
-    "        data.drop(fail, inplace=True)\n",
-    "\n",
-    "        # cut on lepton type using the function cut_lep_type defined above\n",
-    "        fail = data[ np.vectorize(cut_lep_type)(data.lep_type) ].index\n",
-    "        data.drop(fail, inplace=True)\n",
-    "\n",
-    "        # calculation of 4-lepton invariant mass using the function calc_mllll defined above\n",
-    "        data['mllll'] = np.vectorize(calc_mllll)(data.lep_pt,data.lep_eta,data.lep_phi,data.lep_E)\n",
-    "\n",
-    "        # dataframe contents can be printed at any stage like this\n",
-    "        #print(data)\n",
-    "\n",
-    "        # dataframe column can be printed at any stage like this\n",
-    "        #print(data['lep_pt'])\n",
-    "\n",
-    "        # multiple dataframe columns can be printed at any stage like this\n",
-    "        #print(data[['lep_pt','lep_eta']])\n",
-    "\n",
-    "        nOut = len(data.index) # number of events passing cuts in this batch\n",
-    "        data_all = data_all.append(data) # append dataframe from this batch to the dataframe for the whole sample\n",
-    "        elapsed = time.time() - start # time taken to process\n",
-    "        print(\"\\t\\t nIn: \"+str(nIn)+\",\\t nOut: \\t\"+str(nOut)+\"\\t in \"+str(round(elapsed,1))+\"s\") # events before and after\n",
+    "    data_all = [] # define empty list to hold all data for this sample\n",
+    "    \n",
+    "    # open the tree called mini using a context manager (will automatically close files/resources)\n",
+    "    with uproot.open(path + \":mini\") as tree:\n",
+    "        numevents = tree.num_entries # number of events\n",
+    "        if 'data' not in sample: xsec_weight = get_xsec_weight(sample) # get cross-section weight\n",
+    "        for data in tree.iterate(['lep_pt','lep_eta','lep_phi',\n",
+    "                                  'lep_E','lep_charge','lep_type', \n",
+    "                                  # add more variables here if you make cuts on them \n",
+    "                                  'mcWeight','scaleFactor_PILEUP',\n",
+    "                                  'scaleFactor_ELE','scaleFactor_MUON',\n",
+    "                                  'scaleFactor_LepTRIGGER'], # variables to calculate Monte Carlo weight\n",
+    "                                 library=\"ak\", # choose output type as awkward array\n",
```

Comment:
change output type to awkward array (note the slightly different syntax for uproot4)